specs/pvs/corla.pvs

```diff
@@ -1650,6 +1761,186 @@ END system_assumptions
+
+%%% \section{Data Model}
+
+%%% A single, mirrored relational database is used to store all
+%%% persistent information for the RLA Tool.  In order to define the
+%%% data model of the system, we need to:
+%%% \begin{enumerate}
+%%% \item identify the \emph{core concepts} that must be persistent in
+%%% the system
+%%% \item identify the \emph{relationships} that must be persistent in the
+%%% system
+%%% \item derive from these pieces of information:
+%%%   \begin{enumerate}
+%%%   \item the necessary \emph{tables} that \emph{collect concepts},
+%%%   \item identify the \emph{indexing concept} of each table, and
+%%%   \item define the appropriate \emph{joins} that realize
+%%%         relationships
+%%%   \end{enumerate}
+%%% \end{enumerate}
+
+data_model: THEORY
+BEGIN
+  IMPORTING fafessl, database, information_systems, roles, rla_tool,
+            authentication, rlas, state_wide_dashboard,
+            cast_vote_records
+
+  be: VAR back_end
+  db: VAR database
+
+%%% The first set of data persistently stored relates to
+%%% authentication. Some of this data will be stored in the two-factor
+%%% authentication system provided by CDOS. All other data must be
+%%% stored in the RLA Tool database. In general, each entry in the
+%%% authentication table is simply an administrator and two-factor
+%%% authentication credential pair.
+  authentication_table: TYPE =
+    sequence[[administrator, two_factor_authentication]]
+
+%%% From the State's point of view, the State-wide dashboard permits
+%%% State administrators to establish risk limits for each contest,
+%%% specify which kinds of audits are used for which races and in
+%%% which counties, etc.  Most of the dynamic data available to State
+%%% administrators is provided via County administrators using the
+%%% County dashboard.
+
+  state_status_table: TYPE =
+    sequence[[# county: county,
+                general_status: county_status,
+                audit_status: audit_stage,
+                date: string #]]
+
+%%% There is also some background information that relates to the RLA
+%%% overall, much of which is relevant to the State-wide dashboard. In
+%%% particular, geographic (such as the identity of the State and its
+%%% Counties) and political (registered political parties) information
+%%% must be stored, as must information about each election under
+%%% audit.
+%%%
+%%% Notice that these tables are simply a flattening of the election
+%%% datatype. This is a standard pattern in mapping HOL specifications
+%%% to relational data models.
+  geography_table: TYPE =
+    sequence[[nation, state, county]]
+  political_party_table: TYPE =
+    sequence[[political_party]]
+  national_contest_table: TYPE =
+    sequence[[nation, contest]]
+  state_contest_table: TYPE =
+    sequence[[state, contest]]
+  county_contest_table: TYPE =
+    sequence[[county, contest]]
+
+  public_meeting_to_determine_seed_table: TYPE =
+    [# location: string, date: string #]
```

Comment:
I'll just note as I did re: CollectionID vs CountyID above, that there will be times where counties will want to add more ballots after the random seed is selected. If they want to include those in an audit, there will then have to be another random seed generation ceremony.  But hopefully not in November 2017, since we don't have well-documented algorithms (e.g. stratified Bayes audit) to specifically deal with it yet....